### PR TITLE
Support Packetbeat in Agent

### PR DESCRIPTION
**Pull request by andrewstucki**
## What does this PR do?

This is a first pass at setting up packetbeat for use with agent. Some things that it does:

1. Refactors the build tooling of packetbeat to use the same build tooling as filebeat/metricbeat/heartbeat for the sake of integrating with the agent build/packaging scripts
2. Adds a (somewhat) sane configuration agent configuration mapping for packetbeat that filters for input types of `network/protocols/*` or `network/flows`
3. Adds a new AST transformation rule to handle copying an entry from an array into a single dictionary entry, this is necessary to move `type: network/flow` into the singleton configuration entry `packetbeat.flows`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes https://github.com/elastic/beats/issues/21356

### CODE REVIEW

File: x-pack/elastic-agent/pkg/agent/program/testdata/single_config.yml [hunk 1]
```diff
@@ -25,145 +25,151 @@ outputs:
...
-        policy: metrics-endpoint.policy-default
-        telemetry: metrics-endpoint.telemetry-default
-      logging:
-        file: info
-        stdout: debug
-      events:
-        process: true
-        file: true
-        network: true
+  - type: docker/metrics
```
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/program/testdata/single_config.yml:28]: We should probably "standardise" on the indentation to not have too big diffs here in the future. Linting rule perhaps?

**Review by ruflin — COMMENTED** [review]
@andrewstucki Great to see this PR.

Could you move your first commits around the refactoring of packetbeat build process into a separate PR? Like this it can be reviewed / tested / merged separately and hopefully get it in quickly.

For the transpiler, the long term goal is to mostly get rid of it. It means, the processes should understand the config language of the Agent instead of the other way around. We are doing this for heartbeat and apm-server, hopefully the other Beats too in the future. Now the questions is what we should do with packetbeat and how tricky it is to add the new inputs format support to packetbeat instead.

Could you share some yaml examples on how the packetbeat config will look like in the Agent with before / after? I see there is one for amqp but would be good to have a few more examples for potential discussion.

An other topic we should discuss if packetbeat should be packaged with the Agent by default or not.

**Comment by andrewstucki** [comment]
@ruflin - sounds good -- I'll split out the build configuration right away and open up another PR. For the packetbeat config my plan was to pretty much just to do whatever we needed for splitting out the flows configuration, and then have an almost one to one of `inputs` to the values from `packetbeat.protocols` (just take a look under the `packetbeat.protocols` section in `packetbeat.reference.yml` and you'll get an idea). For any additional global configuration (i.e. interface devices, etc.) the thought was to leverage defaults for now (i.e. sniff on all devices). I would imagine that apart from doing some of the configuration munging required for splitting out the flow configuration, moving the logic into packetbeat itself wouldn't be too bad.

**Comment by ruflin** [comment]
Thanks for splitting it out.

I agree, the config will probably be straight forward. TBH I also ask for examples to have a place where I can send people to not aware of all the details and see examples. So I think it would still be good to have one written out for 3 inputs (aka protocols) + flow.

If you keep the structure and naming of each protocol, agree it should be pretty straight forward and on our end, everthing prefixed with network we forward to packetbeat process.

I wonder if you need the protocol prefix part or could just go with `network/dhcpv4` as an example?

**Pull request closed without merging**